File: conformance/runner/typescript/case-census.ts
/**
 * Case census (#602): every non-live fixture case must be accounted for by the
 * run.
 *
 *     registered cases  ==  every JSON case under conformance/tests,
 *                           recursively, whose mode != "live"
 *
 * The left side is what the runner actually registered with vitest. The right
 * side is counted by `countNonLiveCases` below — a SEPARATE walk and parse,
 * deliberately not the runner's own load path. That independence is the entire
 * point: a check fed by the load path can only confirm the load path agrees
 * with itself.
 *
 * Why `mode !== "live"` rather than `mode === "mock"`: all six runners select
 * with "mock unless told otherwise" (`isMockMode` here, and its five
 * equivalents), so a typo'd `mode: "moc"` is dropped by every runner at once
 * with nothing printed anywhere. Counting the expected side as "not explicitly
 * live" turns that silent divergence into arithmetic.
 *
 * Catches: an unrecognized `mode`; a fixture that failed to parse or was never
 * globbed (including one nested below `conformance/tests/`, which no runner
 * discovers — hence the recursive walk); a case dropped between load and
 * dispatch; a fixture emptied to `[]` (which the census REFUSES rather than
 * counts — see `countNonLiveCases`, and note that counting it would make this
 * bullet a lie); and any future skip channel that bypasses registration,
 * because registration is what it reads.
 *
 * The typo is not this check's alone to catch, and saying so is what keeps the
 * rest of the list honest: `make conformance-fixtures-check` validates
 * `conformance/tests/*.json` against `conformance/schema.json`, whose `mode` is
 * `enum: ["mock", "live"]`, so a typo in a TOP-LEVEL fixture fails there first
 * and this census is defense in depth for that one case. What that gate
 * structurally cannot see is everything else above — its glob is not recursive,
 * so a fixture nested below `conformance/tests/` is validated by nothing AND run
 * by nothing (verified: such a file passes the schema gate and fails this
 * census); a fixture truncated to `[]` is a valid array of zero cases; and a
 * case dropped between load and registration is not a fixture-format question at
 * all. Nor does that gate run when `make conformance-typescript` is invoked
 * alone.
 *
 * Does NOT catch the all-six case #602 names — one case every runner excludes
 * for its own reason, which leaves each runner's own census green. That needs
 * the six exclusion sets in one place, hence artifact plumbing across six CI
 * jobs; #602 stays open for it.
 *
 * TYPESCRIPT COUNTS REGISTRATIONS, NOT OUTCOMES, and that is deliberate. vitest
 * owns pass/fail/skip accounting and `it.skip` is how this runner expresses a
 * skip, so there are no counters to sum the way the other five runners do.
 * Asserting on the number of registered cases is the same invariant expressed
 * where vitest can see it: a case dropped at load never becomes an `it` at all.
 * The TS lane is also the one with no `SKIP: <name>` line, so a dropped case
 * leaves no trace in its output either.
 */

import * as fs from "node:fs";
import * as path from "node:path";

/**
 * Whether a fixture case's `mode` selects the mock runner.
 *
 * Absent means mock: live cases belong to live-runner.test.ts (the canonical
 * wire-capturer), and every other value is nobody's. Shared with the census
 * self-tests so the rule the load path applies is the rule under test, not a
 * copy of it.
 */
export function isMockMode(mode: string | undefined): boolean {
  return (mode ?? "mock") === "mock";
}

/** Every `*.json` file under `dir`, recursively, sorted by path. */
function fixtureFiles(dir: string): string[] {
  const found: string[] = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      found.push(...fixtureFiles(full));
    } else if (entry.isFile() && entry.name.endsWith(".json")) {
      found.push(full);
    }
  }
  return found.sort();
}

/**
 * Counts fixture cases whose mode is not `"live"`, recursively.
 *
 * Fail-closed in three places, each a way the count could certify nothing while
 * looking green: an unreadable tree, a fixture that does not parse, and a walk
 * that found no fixture files at all.
 */
export function countNonLiveCases(testsDir: string): number {
  const files = fixtureFiles(testsDir);
  if (files.length === 0) {
    throw new Error(`no *.json fixture files found under ${testsDir}`);
  }

  let cases = 0;
  for (const file of files) {
    let parsed: unknown;
    try {
      parsed = JSON.parse(fs.readFileSync(file, "utf-8"));
    } catch (err) {
      throw new Error(`${file}: ${err instanceof Error ? err.message : String(err)}`);
    }
    if (!Array.isArray(parsed)) {
      throw new Error(`${file}: fixture is not a JSON array`);
    }
    // An emptied fixture is REFUSED, not counted as zero, and this is the one
    // rejection that carries the whole-file guarantee. It is the single
    // truncation both sides of the census read identically: the runner
    // registers nothing from the file and the census expects nothing, so the
    // two totals fall together and no mismatch ever appears. Counting it would
    // make "a fixture truncated to []" a claim this check cannot keep. A file
    // declaring no cases tests nothing, so refusing it costs nothing — and it
    // closes the same hole in conformance-fixtures-check, where an empty array
    // is a schema-valid list of zero items.
    if (parsed.length === 0) {
      throw new Error(
        `${file}: fixture declares no cases; delete the file or restore its cases`,
      );
    }
    // Only `mode` is read: the census must survive a fixture whose other fields
    // this runner cannot model, or it would report a failure for a case the run
    // itself handled fine.
    cases += parsed.filter((tc) => (tc as { mode?: string })?.mode !== "live").length;
  }
  return cases;
}

/**
 * Compares what the run registered against the census, returning null when they
 * agree and a message naming the short side otherwise.
 */
export function caseCountFailure(registered: number, expected: number): string | null {
  if (registered === expected) return null;
  if (registered < expected) {
    return (
      `case census: the run registered ${registered} case(s) but conformance/tests ` +
      `holds ${expected} non-live case(s) — ${expected - registered} executed by ` +
      "nothing. An unrecognized `mode`, a fixture that failed to parse or was never " +
      "globbed, or a case dropped between load and registration will do this."
    );
  }
  return (
    `case census: the run registered ${registered} case(s) but conformance/tests ` +
    `holds only ${expected} non-live case(s) — ${registered - expected} more than ` +
    "the fixtures declare."
  );
}

File: conformance/runner/go/case_census.go
package main

// Case census (#602): every non-live fixture case must be accounted for by the
// run.
//
//	passed + failed + skipped  ==  cases in conformance/tests/**/*.json
//	                               whose mode != "live"
//
// The left side is what the runner actually did. The right side is counted by
// countNonLiveCases below — a SEPARATE walk and parse, deliberately not the
// runner's own load path. That independence is the entire point: a check fed by
// the load path can only ever confirm the load path agrees with itself.
//
// Why `mode != "live"` rather than `mode == "mock"`. All six runners select
// cases with "mock unless told otherwise" (`isMockMode` here, and its five
// equivalents), so a typo'd `mode: "moc"` is dropped by every runner at once
// with nothing printed anywhere. Counting the expected side as "not explicitly
// live" is what turns that silent divergence into arithmetic.
//
// What it catches: a typo'd or otherwise unrecognized `mode`; a fixture file
// that failed to parse or was never globbed (including one nested below
// conformance/tests/, which no runner discovers — hence the recursive walk); a
// case dropped between load and dispatch; a whole fixture emptied to `[]` (which
// the census REFUSES rather than counts — see countNonLiveCases, and note that
// counting it would make this bullet a lie); and any future skip channel that
// bypasses the counters, because the counters are what it reads rather than any
// particular skip mechanism.
//
// The typo is not this check's alone to catch, and saying so is what keeps the
// rest of the list honest: `make conformance-fixtures-check` validates
// conformance/tests/*.json against conformance/schema.json, whose `mode` is
// `enum: ["mock", "live"]`, so a typo in a TOP-LEVEL fixture fails there first
// and this census is defense in depth for that one case. What that gate
// structurally cannot see is everything else above — its glob is not recursive,
// so a fixture nested below conformance/tests/ is validated by nothing AND run
// by nothing (verified: such a file passes the schema gate and fails this
// census); a fixture truncated to `[]` is a valid array of zero cases; and a
// case dropped between load and dispatch is not a fixture-format question at
// all. Nor does that gate run when `make conformance-<lang>` is invoked alone.
//
// What it does NOT catch, stated rather than implied: the literal all-six case
// #602 names — one fixture case that every runner excludes for its own separate
// reason. Each runner's census is green in that situation, because each runner
// counted its own skip. Detecting it needs the six exclusion sets in one place,
// which needs artifact plumbing across six CI jobs; #602 stays open for that.

import (
	"encoding/json"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// isMockMode reports whether a fixture case's `mode` selects this runner.
//
// Absent means mock: live cases are TS-only (the canonical wire-capturer), and
// every other mode value is nobody's. Shared with the census self-tests so the
// rule the run loop applies is the rule under test, not a copy of it.
//
// The parameter is a pointer because a plain string cannot tell an ABSENT key
// from `"mode": ""`, and the two must not be the same answer. The other five
// runners null-coalesce, so they run the first and refuse the second; a Go
// `string` collapses both to "" and would run an unrecognized mode the other
// five reject — a hole in the very invariant this census exists to hold, in the
// reference runner.
func isMockMode(mode *string) bool {
	return mode == nil || *mode == "mock"
}

// countNonLiveCases counts the fixture cases whose mode is not "live",
// recursively, under dir.
//
// Fail-closed in three places, each of which is a way the count could certify
// nothing while looking green: an unreadable tree, a fixture that does not
// parse, and a walk that found no fixture files at all.
func countNonLiveCases(dir string) (int, error) {
	cases, files := 0, 0

	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() || !strings.HasSuffix(d.Name(), ".json") {
			return nil
		}
		files++

		raw, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		// Only `mode` is read: the census must survive a fixture whose other
		// fields this runner cannot model, or it would report a parse failure
		// for a case the run itself handled fine.
		var parsed []struct {
			Mode *string `json:"mode"`
		}
		if err := json.Unmarshal(raw, &parsed); err != nil {
			return fmt.Errorf("%s: %w", path, err)
		}
		// A top-level `null` unmarshals into a nil slice without error, so it
		// would otherwise pass as a fixture of zero cases — the one non-array
		// root the other five runners reject and this one would not.
		if parsed == nil {
			return fmt.Errorf("%s: fixture is not a JSON array of cases", path)
		}
		// An emptied fixture is REFUSED, not counted as zero, and this is the
		// one rejection that carries the whole-file guarantee. It is the single
		// truncation both sides of the census read identically: the runner
		// registers nothing from the file and the census expects nothing, so
		// the two totals fall together and no mismatch ever appears. Counting
		// it would make "a fixture truncated to []" a claim this check cannot
		// keep. A file declaring no cases tests nothing, so refusing it costs
		// nothing — and it closes the same hole in conformance-fixtures-check,
		// where an empty array is a schema-valid list of zero items.
		if len(parsed) == 0 {
			return fmt.Errorf("%s: fixture declares no cases; delete the file or restore its cases", path)
		}
		for _, c := range parsed {
			if c.Mode == nil || *c.Mode != "live" {
				cases++
			}
		}
		return nil
	})
	if err != nil {
		return 0, err
	}
	if files == 0 {
		return 0, fmt.Errorf("no *.json fixture files found under %s", dir)
	}
	return cases, nil
}

// caseCountFailure compares what the run accounted for against the census,
// returning "" when they agree and a message naming the short side otherwise.
func caseCountFailure(ran, expected int) string {
	switch {
	case ran == expected:
		return ""
	case ran < expected:
		return fmt.Sprintf(
			"case census: the run accounted for %d case(s) (passed+failed+skipped) "+
				"but conformance/tests holds %d non-live case(s) — %d executed by nothing. "+
				"An unrecognized `mode`, a fixture that failed to parse or was never globbed, "+
				"or a case dropped between load and dispatch will do this.",
			ran, expected, expected-ran)
	default:
		return fmt.Sprintf(
			"case census: the run accounted for %d case(s) (passed+failed+skipped) "+
				"but conformance/tests holds only %d non-live case(s) — %d more than the fixtures declare.",
			ran, expected, ran-expected)
	}
}
